### DISCUSSION TITLE
Adjusting the grid snap tile placement for the stamp brush

### DISCUSSION THREAD

**Discussion by kadattack** [mapeditor/tiled#3886] [Q&A] [answered]
Can anyone point me to the file that contains the code that decides what grid size the Stamp Brush places tiles on the Map?
I want to change the code so that it will always paste the tiles only every 32x32 box regardless of grid size.

Here is a screenshot with 2 column examples, on the left, to explain what i mean. Currently i have four 16x16 tiles selected together (as shown in the screenshot) and if i drag the mouse downward, the brown dirt tile will stamp over green wall tile and you will end up with the wall showing only at the end of the mouse stroke. However what i actually want is the right column example where, when i drag downwards, it will only paste once every 32x32 even though the grid size is 16x16.
![Screenshot from 2024-02-10 14-31-51](https://github.com/mapeditor/tiled/assets/78758877/f4f192fc-1759-4d7e-b917-796f87be26f1)

The reason why i need this is too complex for me to explain, but basically the issue comes with Automap placement and me using 4 dirt tiles that don't have the same index.

**Comment by kadattack** [comment, marked as answer]
Ok I found out how to do it.

I changed the functions 
[void StampBrush::tilePositionChanged(QPoint pos)
](https://github.com/mapeditor/tiled/blob/0e3dfd624dbffb84d2c3540034f65b9e4e936f9c/src/tiled/stampbrush.cpp#L86)
and
[void StampBrush::updatePreview()
](https://github.com/mapeditor/tiled/blob/0e3dfd624dbffb84d2c3540034f65b9e4e936f9c/src/tiled/stampbrush.cpp#L600)
so they only work when QPoint coordiantes are even numbers.

**Comment by eishiya** [comment]
For folks with similar needs but who don't want to go as far as modifying Tiled, or who might need to use a mix of 2x2 and regular placement:
If you hold Shift, you can draw lines with the stamp, which will draw in appropriately-sized blocks, drawing correct rows and columns with no overlap. So, as long as your stroke starts in a correct place (which you can check in the status bar or visually via the major grid), it'll match the crid.
For Automapping, you can use multiple input layers to react to any of the tiles rather than to just a specific one, so perfectly precise input is not required. You can also use Automapping's offsetX/Y and modX/Y properties to create rules that follow a specific (e.g. 2x2) grid without your source tiles having to be in such a grid.
